### PR TITLE
Fix Illegal Instruction/IMA errors when using DP attention with DeepSeek-V3.2 models

### PR DESCRIPTION
**Pull request by YAMY1234**
## Motivation

Fix illegal memory access (IMA) errors when using DP attention with DeepSeek-V3.2 models.

**Issue**: When running with `--enable-dp-attention`, the DP gather operation in logits stage uses incorrect metadata, causing Triton kernel to access out-of-bounds memory.

**Root cause**: The `get_dp_local_info()` function always uses `global_num_tokens_gpu` (total tokens for attention/MLP), but in logits stage it should use `global_num_tokens_for_logprob_gpu` (pruned tokens needing logits computation).

### 1. Fix metadata source selection in `get_dp_local_info()`

**File**: `python/sglang/srt/layers/dp_attention.py`

Added logic to distinguish between two calling contexts:
- **LogitsMetadata**: Use `global_num_tokens_for_logprob_gpu` (correct for logits stage)
- **ForwardBatch**: Use `global_num_tokens_gpu` (correct for attention/MLP stage)

This ensures offset calculation uses the appropriate metadata field.

### 2. Use actual tensor size in `_dp_gather_via_all_reduce()`

**File**: `python/sglang/srt/layers/dp_attention.py`

Override `local_num_tokens` with `local_tokens.shape[0]` before calling `memcpy_triton`. Uses in-place `.fill_()` to maintain CUDA graph compatibility.

This provides a safety net since scheduler's `num_tokens_for_logprob` calculation doesn't account for logits_processor pruning.

## Accuracy Tests

```
python -m sglang.launch_server --model deepseek-ai/DeepSeek-V3.2-Exp --tp 8 --dp 8 --enable-dp-attention

python3 -m sglang.test.run_eval --port 30000 --eval-name gpqa --num-examples 198 --max-tokens 36000 --repeat 8 --thinking-mode deepseek-v3
```

```
Scores: ['0.798', '0.803', '0.803', '0.798', '0.788', '0.778', '0.773', '0.798']█████████████████████████████████████████████████████████████████████████████████▉      | 185/198 [41:56<01:16, 5.85s/it]
====================
Writing report to /tmp/gpqa_deepseek-ai_DeepSeek-V3.2-Exp.html███████████████▌                                                | 85/198 [42:52<28:55, 15.36s/it]
{'chars': np.float64(14306.5), 'chars:std': np.float64(11493.867823449567), 'score:std': np.float64(0.4015072103909452), 'score': np.float64(0.797979797979798)}
Writing results to /tmp/gpqa_deepseek-ai_DeepSeek-V3.2-Exp.json
Total latency: 2573.581 s
Score: 0.798
```

```
Repeat: 8, mean: 0.799
Scores: ['0.793', '0.798', '0.813', '0.808', '0.813', '0.783', '0.768', '0.818']
====================
Writing report to /tmp/gpqa_deepseek-ai_DeepSeek-V3.2-Exp.html
{'chars': np.float64(14737.838383838383), 'chars:std': np.float64(11840.990410141889), 'score:std': np.float64(0.38569460791993493), 'score': np.float64(0.8181818181818182)}
Writing results to /tmp/gpqa_deepseek-ai_DeepSeek-V3.2-Exp.json
Total latency: 1584.823 s
Score: 0.818
```

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

**Comment by YAMY1234** [comment]
Resolving https://github.com/sgl-project/sglang/issues/11942

File: python/sglang/srt/layers/dp_attention.py [hunk 1]
```diff
@@ -355,16 +355,27 @@ def disable_dp_size():
...
-    # `get_dp_local_info` is only called in global DP gather and scatter. We use global DP rank here.
+    # Compute local token range for current DP rank.
+    # For LogitsMetadata, use global_num_tokens_for_logprob_gpu (pruned tokens needing logits).
+    # For ForwardBatch (MLP/Attention), use global_num_tokens_gpu (all tokens in batch).
     dp_rank = get_attention_dp_rank()
 
     if forward_batch.dp_local_start_pos is None:
-        cumtokens = torch.cumsum(forward_batch.global_num_tokens_gpu, dim=0)
+        # Select metadata source based on context
+        if (
```
Comment (ch-wan) [inline python/sglang/srt/layers/dp_attention.py]: How can `dp_local_start_pos` be None in `LogitsMetadata`? It is expected to be processed here: https://github.com/sgl-project/sglang/blob/0bfa394affbf62db60390411294ae6c67cb24455/python/sglang/srt/layers/logits_processor.py#L577.

`LogitsMetadata` has a separate processing logic. We should keep it separated. Perhaps we should add assertion like `assert isinstance(forward_batch, ForwardBatch)` to avoid misuse.

Comment (YAMY1234) [reply python/sglang/srt/layers/dp_attention.py]: Thanks, understand the design! The call logic should indeed stay separated at the outer level. Will have another patch soon after verified

Comment (YAMY1234) [reply python/sglang/srt/layers/dp_attention.py]: @ch-wan Just pushed another patch, could you take another look? Thanks

**Comment by YAMY1234** [comment]
Root fix is now in a new PR: https://github.com/sgl-project/sglang/pull/12115 which can resolve the issue with in a more robust way. Closing this PR
